chinese_tutorial_rvest_httr.Rmd

```diff
@@ -0,0 +1,490 @@
+| 所有节点                              | `~`                        | `*`              | `//*`                         |
+| 查找一级、二级、三级标题节点          | `<h1>`,`<h2>`,`<h3>`       | `h1`,`h2`,`h3`   | `//h1`,`//h2`,`//h3`          |
+| 所有的 P 节点                         | `<p>`                      | `p`              | `//p`                         |
+| p 节点的所有子节点                    | `<p>`标签下的所有节点      | `p > *`          | `//p/*`                       |
+| 查找所有包含 attr 属性的 li 标签      | `<li attr="~">`            | `li[attr]`       | `li[@attr]`                   |
+| 查找所有 attr 值为 value 的 li 标签   | `<li attr="value">`        | `li[attr=value]` | `//li[@attr='value']`         |
+| 查找 id 值为 item 的所有 div 节点     | `<div id="item">`          | `div#item`       | `//div[@id='item']`           |
+| 查找 class 值中包含 foo 的所有标签    | `<* class="foo blahblah">` | `.foo`           | `//*[contains(@class,'foo')]` |
+| 第一个 P 节点                         | 所有`<p>`中的第一个 `<p>`  | `p:first-child`  | `//p[1]`                      |
+| 第 n 个 P 节点                        | 所有`<p>`中的第 n 个 `<p>` | `p:nth-child(n)` | `//p[n]`                      |
+| 拥有子节点 a 的所有 P 节点            | `<p><a></p>`               | css 无法实现     | `//p[a]`                      |
+| 查找文本内容是“Web Scraping”的 p 节点 | ``<p>`Web Scraping`</p>``  | css 无法实现     | `//p[text()="Web Scraping"]`  |
+
+关于html, css, xpath详细内容可以参考https://www.w3schools.com/html/default.asp， 在此不再赘述。
+
+### 案例一：抓取 CRAN 上所有 R 包的信息
+
+链接：https://cran.rstudio.com/web/packages/available_packages_by_name.html
+
+要求：获取 CRAN 上所有 R 包的信息（名称、简介）
+
+1. 分析网页结构
+2. 抓取网页数据
+
+```R
+u <- "https://cran.rstudio.com/web/packages/available_packages_by_name.html"
+page <- u %>% read_html()
+pkg_table <- page %>% html_node('table') %>% html_table(fill=TRUE) 
+str(pkg_table)
+```
+
+```R
+## 'data.frame':    11063 obs. of  2 variables:
+##  $ X1: chr  "" "A3" "abbyyR" "abc" ...
+##  $ X2: chr  NA "Accurate, Adaptable, and Accessible Error Metrics for Predictive\nModels" "Access to Abbyy Optical Character Recognition (OCR) API" "Tools for Approximate Bayesian Computation (ABC)" ...
+#View(pkg_table)
+```
+
+3. 数据的变形、清洗
+
+前面我们已经看到表格的第一行存在缺失值，那么我们接下来就要对表格进行数据清洗:
+
+```R
+# 删除缺失值
+pkg_table <- pkg_table[complete.cases(pkg_table),]
+# 定义表头
+colnames(pkg_table) <- c("name","title")
+head(pkg_table,3)
+```
+
+```R
+##     name
+## 2     A3
+## 3 abbyyR
+## 4    abc
+##                                                                      title
+## 2 Accurate, Adaptable, and Accessible Error Metrics for Predictive\nModels
+## 3                  Access to Abbyy Optical Character Recognition (OCR) API
+## 4                         Tools for Approximate Bayesian Computation (ABC)
+```
+
+4. 数据的存储
+
+数据量较小，可以直接写入本地文件。
+
+```R
+# 1. 直接存为 Rdata
+save(pkg_table,file="pkg_table.Rdata")
+# 2. 存为 CSV 文件
+write.table(pkg_table, file = "pkg_table.csv", quote = FALSE,
```

Comment:
The file should not contain any install.packages(), write functions, setwd(), or getwd()

Reply:
I think maybe for translation articles, it's ok to leave functions like this.